modules/video/src/video-builder.ts

```diff
@@ -0,0 +1,75 @@
+const VIDEO_BUILDER_OPTIONS = {
+  width: 512,
+  height: 512,
+  framerate: 24,
+  bitrate: 200,
+  realtime: false
+};
+
+type VideoBuilderOptions = Partial<typeof VIDEO_BUILDER_OPTIONS>;
+
+function nextEvent(target, name): Promise<{data: any}> {
+  return new Promise((resolve) => {
+    target.addEventListener(name, resolve, {once: true});
+  });
+}
+
+export default class VideoBuilder {
+  worker: Worker = new Worker('node_modules/webm-wasm/dist/webm-worker.js');
+
+  static get properties() {
+    return {
+      id: 'webm',
+      name: 'WEBM',
+      extensions: ['webm'],
+      mimeTypes: ['video/webm'],
+      builder: VideoBuilder,
+      options: VIDEO_BUILDER_OPTIONS
+    };
+  }
+
+  async initialize(options: VideoBuilderOptions): Promise<void> {
+    const {bitrate, framerate, height, width, realtime} = {...VIDEO_BUILDER_OPTIONS, ...options};
+
+    this.worker.postMessage('./webm-wasm.wasm');
+    await nextEvent(this.worker, 'message');
+
+    this.worker.postMessage({timebaseDen: framerate, width, height, bitrate, realtime});
+  }
+
+  addFrame(source): void {
+    let buffer = source;
+
+    if (Buffer.isBuffer(source)) {
+      buffer = source;
+    } else if (source instanceof ImageData) {
+      buffer = source.data;
+    } else if (source instanceof HTMLCanvasElement) {
+      buffer = source.getContext('2d')!.getImageData(0, 0, source.width, source.height).data;
+    } else if (source instanceof CanvasRenderingContext2D) {
+      buffer = source.getImageData(0, 0, source.canvas.width, source.canvas.height).data;
+    } else if (
+      source instanceof WebGLRenderingContext ||
+      source instanceof WebGL2RenderingContext
+    ) {
+      const gl = source;
+      // const {width, height} = gl.canvas;
```

Comment:
Either one is fine with me. Preference?

Reply:
Either should be fine if they're both the buffer size. I've assumed at this point `const {width, height} = gl.canvas;` was the way to do it, but I only have limited theoretical knowledge of how it differs from `GL.VIEWPORT`. 

Do you know if `GL.VIEWPORT` is meaningfully different? The [spec](https://www.khronos.org/registry/webgl/specs/latest/1.0/#2.3) suggests they are synced on the application level.

Pulling the correct size has been a source of issues in the past due to subtle differences in how libs handle it. Luma.gl uses [`gl.canvas`](https://github.com/visgl/luma.gl/blob/c3e42ceb635ef0a698f742f27a0c059b3a0244c0/modules/webgl/src/context/context/device-pixels.ts#L80-L81). Mapbox did something different. Maplibrae has diverged now I think. I tried to sort it out in this [issue](https://github.com/visgl/hubble.gl/issues/159) and [rfc](https://github.com/visgl/hubble.gl/pull/176/files), but I wouldn't be surprised if it comes up again in some corner case.

I say go with what you know. I only know enough to flag it might not be straight forward.

Reply:
1. loaders.gl is generally WebGL agnostic - it doesn't depend on WebGL but works with typed arrays which are optimized for WebGL use.
2. Also loaders.gl doesn't use Node.js constructs like `Buffer`, preferring to use typed arrays in all interfaces. Simply mentioning `Buffer` tends to pull in a 50KB polyfill in browser bundles...

Reply:
I would have liked it more if this function just accepted a typed array and if necessary we offered some utility function for getting hold of a typed array from different input types.